### PR TITLE
support for safe timers, periodic scheduling, #16742

### PR DESCRIPTION
**Pull request by patriknw**
Plenty of FIXME but worth early review.

Refs #16742

### CODE REVIEW

File: akka-typed-tests/src/test/scala/akka/typed/TimerSpec.scala [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+            timer.cancel("T")
+            Same
+          case Throw(e) ⇒
+            throw e
+        }
+      }
+
+    def `must schedule repeated ticks`(): Unit = {
+      val probe = TestProbe[Event]("evt")
+      val behv = timerScheduler[Command] { timer ⇒
```
Comment (johanandren) [inline akka-typed-tests/src/test/scala/akka/typed/TimerSpec.scala]: The other behaviours is something the actor is `deferred`, `immutable`, `mutable` while this is something it gets injected. Should it perhaps be `timed` instead, or `withTimer`?

Comment (patriknw) [inline akka-typed-tests/src/test/scala/akka/typed/TimerSpec.scala]: I like `withTimer`

**Comment by patriknw** [comment]
Pushed new implementation that is not based on spawnAdapter.
* spawnAdapter function is invoked synchronously on caller's thread,
  so it can't be used for this
* implemented an Intercept behavior that can handle the
  scheduled TimerMsg, even though it's of a different type
* Intercept is similar to Tap but more powerful, implemented
  Tap with Intercept. Intercept is internal API.

**Comment by patriknw** [comment]
I think this is complete now, ready for final review.

File: akka-typed/src/main/scala/akka/typed/internal/TimerSchedulerImpl.scala [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+import scala.reflect.ClassTag
+
+/**
+ * INTERNAL API
+ */
+@InternalApi private[akka] object TimerSchedulerImpl {
+  final case class Timer[T](key: Any, msg: T, repeat: Boolean, generation: Int, task: Cancellable)
+  final case class TimerMsg(key: Any, generation: Int, owner: AnyRef)
+
+  def withTimer[T](factory: TimerSchedulerImpl[T] ⇒ Behavior[T]): Behavior[T] = {
```
Comment (ktoso) [inline akka-typed/src/main/scala/akka/typed/internal/TimerSchedulerImpl.scala]: Ah, sneaky - very good type juggling so we just need to implement it once but is exposed using the right types to each of the DSLs. I also like the `withXxx` style for all kinds of additional features an actor may want 👍

File: akka-typed/src/main/scala/akka/typed/internal/TimerSchedulerImpl.scala [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+
+  def intercept(behavior: Behavior[T]): Behavior[T] = {
+    // The scheduled TimerMsg is intercepted to guard against old messages enqueued
+    // in mailbox before timer was canceled.
+    // Intercept some signals to cancel timers when when restarting and stopping.
+    BehaviorImpl.Intercept[T, TimerMsg](
+      beforeOnMessage = interceptTimerMsg,
+      beforeOnSignal = (ctx, sig) ⇒ {
+      sig match {
+        case PreRestart | PostStop ⇒ cancelAll()
```
Comment (ktoso) [inline akka-typed/src/main/scala/akka/typed/internal/TimerSchedulerImpl.scala]: Hah! Very nice, biggest improvements of the entire PR to timers in actors in general I think :)

File: akka-typed/src/main/scala/akka/typed/internal/TimerSchedulerImpl.scala [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+    log.debug("Start timer [{}] with generation [{}]", key, nextGen)
+    timers = timers.updated(key, nextTimer)
+  }
+
+  override def isTimerActive(key: Any): Boolean =
+    timers.contains(key)
+
+  override def cancel(key: Any): Unit = {
+    timers.get(key) match {
+      case None    ⇒
```
Comment (ktoso) [inline akka-typed/src/main/scala/akka/typed/internal/TimerSchedulerImpl.scala]: comment that `// do nothing` is intentional?

Comment (ktoso) [inline akka-typed/src/main/scala/akka/typed/internal/TimerSchedulerImpl.scala]: or perhaps it should log when trying to cancel already cancelled one actually?

File: akka-typed-tests/src/test/java/akka/typed/javadsl/ActorCompile.java [hunk 5]
```diff
@@ -56,6 +59,13 @@ public MyMsgB(String greeting) {
     });
   }
 
+  {
+    Behavior<MyMsg> b = Actor.withTimer(timers -> {
```
Comment (ktoso) [inline akka-typed-tests/src/test/java/akka/typed/javadsl/ActorCompile.java]: naming nitpick: perhaps this is `withTimers` as it does not start a timer, but give you the capability to do so?

File: akka-typed-tests/src/test/scala/akka/typed/TimerSpec.scala [hunk 6]
```diff
@@ -0,0 +1,207 @@
...
+    implicit def system: ActorSystem[TypedSpec.Command]
+    implicit val testSettings = TestKitSettings(system)
+
+    val interval = 1.second
+    val dilatedInterval = interval.dilated
+
+    def target(monitor: ActorRef[Event], timer: TimerScheduler[Command], bumpCount: Int): Behavior[Command] = {
+      def bump(): Behavior[Command] = {
+        val nextCount = bumpCount + 1
+        timer.startTimer("T", Tick(nextCount), interval, repeat = true)
```
Comment (ktoso) [inline akka-typed-tests/src/test/scala/akka/typed/TimerSpec.scala]: the API with `repeat = true` looks good in scala however in java it's IMHO a bit confusing to have a boolean in there - always one more guessing step what the boolean means. Sure we don't want to have startPeriodicTimer or other ways to have a dedicated name for them?

Maybe `timer.scheduleOnce` / `timer.schedulePeriodic`? The issue with current scheduler names is that "`schedule`" is periodic, but does not immediately look like it is, so sometimes people do that and don't realise they did a periodic one I thought?

File: akka-typed/src/main/scala/akka/typed/scaladsl/TimerScheduler.scala [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+package akka.typed.scaladsl
+
+import scala.concurrent.duration.FiniteDuration
+
+/**
+ * Support for scheduled `self` messages in an actor.
+ * It is used with `Actor.withTimer`, which also takes care of the
+ * lifecycle of the timers such as cancelling them when the actor
+ * is restarted or stopped.
+ */
```
Comment (ktoso) [inline akka-typed/src/main/scala/akka/typed/scaladsl/TimerScheduler.scala]: Rephrase to make it less hidden that on a restart a timer must be scheduled again?

```scala
/**
 * Support for scheduled `self` messages in an actor.
 * It is used with `Actor.withTimer`. 
 * Timers are bound to the lifecycle of the actor that owns it, 
 * and thus are cancelled automatically when it is restarted or stopped.
 */
```

?

File: akka-typed/src/main/scala/akka/typed/scaladsl/TimerScheduler.scala [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+   */
+  def startTimer(key: Any, msg: T, timeout: FiniteDuration, repeat: Boolean): Unit
+
+  /**
+   * Check if a timer with a given `key` is started.
+   */
+  def isTimerActive(key: Any): Boolean
+
+  /**
+   * Cancel a timer with a given `key`. It's guaranteed that a message from the
```
Comment (ktoso) [inline akka-typed/src/main/scala/akka/typed/scaladsl/TimerScheduler.scala]: I'd rephrase the guarantee a bit, people may not be sure what we mean by previous here.

```
Cancel a timer with a given `key`. 
If canceling a timer that was already canceled, or key never was used to start a timer this operation is a no-op.

It is guaranteed that a message from a canceled timer, including its previous incarnation for the same key,
will not be received by the actor, even though the message might already be enqueued in the mailbox
when cancel is called.
```

?

File: akka-typed/src/main/scala/akka/typed/scaladsl/TimerScheduler.scala [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+  /**
+   * Each timer has a key and if a new timer with same key is started
+   * the previous is cancelled and it's guaranteed that a message from the
+   * previous timer is not received, even though it might already be enqueued
+   * in the mailbox when the new timer is started.
+   */
+  def startTimer(key: Any, msg: T, timeout: FiniteDuration, repeat: Boolean): Unit
+
+  /**
+   * Check if a timer with a given `key` is started.
```
Comment (ktoso) [inline akka-typed/src/main/scala/akka/typed/scaladsl/TimerScheduler.scala]: is started => is active

File: akka-typed/src/main/scala/akka/typed/scaladsl/TimerScheduler.scala [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+ * lifecycle of the timers such as cancelling them when the actor
+ * is restarted or stopped.
+ */
+trait TimerScheduler[T] {
+
+  /**
+   * Each timer has a key and if a new timer with same key is started
+   * the previous is cancelled and it's guaranteed that a message from the
+   * previous timer is not received, even though it might already be enqueued
+   * in the mailbox when the new timer is started.
```
Comment (ktoso) [inline akka-typed/src/main/scala/akka/typed/scaladsl/TimerScheduler.scala]: good semantics 👍

File: akka-typed/src/main/scala/akka/typed/scaladsl/Actor.scala [hunk 11]
```diff
@@ -210,6 +211,15 @@ object Actor {
     def wrap[T](b: Behavior[T]): Behavior[T] = akka.typed.internal.Restarter(Behavior.validateAsInitial(b), strategy)(c)
   }
 
+  /**
+   * Support for scheduled `self` messages in an actor.
+   * It takes care of the lifecycle of the timers such as cancelling them when the actor
+   * is restarted or stopped.
```
Comment (ktoso) [inline akka-typed/src/main/scala/akka/typed/scaladsl/Actor.scala]: 👍

**Review by ktoso — APPROVED** [review]
Some small comments, looking great though 👍  
LGTM

File: akka-typed-tests/src/test/java/akka/typed/javadsl/ActorCompile.java (hunk 5 above)
Comment (patriknw) [inline akka-typed-tests/src/test/java/akka/typed/javadsl/ActorCompile.java]: good suggestion

File: akka-typed-tests/src/test/scala/akka/typed/TimerSpec.scala (hunk 6 above)
Comment (patriknw) [inline akka-typed-tests/src/test/scala/akka/typed/TimerSpec.scala]: I like that, changed to `startPeriodicTimer` and `startSingleTimer`

File: akka-typed/src/main/scala/akka/typed/scaladsl/TimerScheduler.scala (hunk 8 above)
Comment (patriknw) [inline akka-typed/src/main/scala/akka/typed/scaladsl/TimerScheduler.scala]: good doc improvements

**Comment by patriknw** [comment]
Pushed the fixes for Deferred and PostStop, and incorporated feedback

**Review by johanandren — COMMENTED** [review]
LGTM
